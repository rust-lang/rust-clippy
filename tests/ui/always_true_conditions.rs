#![warn(clippy::always_true_conditions)]

fn foo_m(name: &str) {
    if name != "Min" && name != "Max" {
        todo!()
    } else {
        todo!()
    }
}

fn foo_s(name: &str) {
    if name != "Min" || name != "Max" {
        //~^ always_true_conditions
        todo!()
    }
    if name != "Min" && name != "Max" {
        todo!()
    }
}

fn catch_or_failure(input: &str) {
    let b = true;
    if b || input != "foo" {
        todo!()
    }
}

fn catch_scope_or_failures(input: &str) {
    let b = true;
    { if b || input != "foo" {} }
}

fn catch_eq_failures() {
    let res = "test";
    if res == "foo" || res == "bar" {
        todo!()
    }
}

fn catch_diff_var_failure(input: &str) {
    let b = "value";
    if b != "bar" || input != "foo" {
        todo!()
    }
}

fn catch_yoda_notation(input: &str) {
    let b = 2;
    if 3 != b || 5 != b {
        //~^ always_true_conditions
        todo!()
    }

    if b != 3 || 5 != b {
        //~^ always_true_conditions
        todo!()
    }
}

fn non_if_funcitonality_tests(input: &str) {
    let x: bool = ("a" != input || "b" != input);
    //~^ always_true_conditions
    let y: bool = ("a" == input || "b" != input);
    let z: bool = (input != "2" || input != "4");
    //~^ always_true_conditions
}
