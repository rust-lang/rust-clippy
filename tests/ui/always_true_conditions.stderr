error: expression will always be true, did you mean to use &&?
  --> tests/ui/always_true_conditions.rs:12:8
   |
LL |     if name != "Min" || name != "Max" {
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::always-true-conditions` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::always_true_conditions)]`

error: expression will always be true, did you mean to use &&?
  --> tests/ui/always_true_conditions.rs:49:8
   |
LL |     if 3 != b || 5 != b {
   |        ^^^^^^^^^^^^^^^^

error: expression will always be true, did you mean to use &&?
  --> tests/ui/always_true_conditions.rs:54:8
   |
LL |     if b != 3 || 5 != b {
   |        ^^^^^^^^^^^^^^^^

error: expression will always be true, did you mean to use &&?
  --> tests/ui/always_true_conditions.rs:61:19
   |
LL |     let x: bool = ("a" != input || "b" != input);
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: expression will always be true, did you mean to use &&?
  --> tests/ui/always_true_conditions.rs:64:19
   |
LL |     let z: bool = (input != "2" || input != "4");
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 5 previous errors

